## **What is an API?**
An API, or Application Programming Interface, is a bridge that allows two different systems to communicate with each other remotely. Consider this analogy:

- You go to a restaurant and place an order with the waiter.
- The waiter conveys your order to the kitchen.
- The kitchen prepares the dish and gives it to the waiter.
- The waiter then serves you the dish.
In this scenario:

- **You (the client)** generate the request.
- **The waiter (the API)** carries the request and the response between you and the kitchen.
- **The kitchen (the server)** processes the request and provides the data or service.
APIs serve as intermediaries, handling requests from clients and fetching or modifying data from servers accordingly. This ensures seamless data exchange across different systems.

## **Key Components in API Communication**
1. **Client:** The device or application initiating the request (e.g., your laptop or smartphone).
2. **API:** Acts as the messenger, transferring requests and responses between client and server.
3. **Server:** The system where data or services reside, responding to client requests.
**Example:** Using ChatGPT involves your device (client) sending a request through an API to ChatGPT’s servers located remotely. The server processes your query and sends back the response via the API.

### **How Do APIs Work?**
1. **Request:** The client sends a request to the API specifying what data or action is required.
2. **Processing:** The server receives the request, processes it, and prepares a response.
3. **Response:** The API sends the processed data or confirmation back to the client.

## **Using APIs in Python**
#### **Request Module**
Python offers a popular package called `requests` (note the plural) to work with APIs, especially REST APIs. It allows developers to generate HTTP requests without loading unnecessary modules.

**Installation:**

> pip install requests

**Common HTTP Methods:**
- **GET:** Retrieve data from a server.
- **POST:** Send data to a server to create or update resources.

### **Example: Making a GET Request**

In [9]:
import requests

# Define the URL of the API endpoint
url = "https://jsonplaceholder.typicode.com/posts"

# Send the GET request
response = requests.get(url)

# Print the status code and content
print(f"Status Code: {response.status_code}")
# response = response[0]

# Get the JSON response
data = response.json()

# Print only the first 10 results
print("First 5 Posts:")
for post in data[:5]:
    print(post)

Status Code: 200
First 5 Posts:
{'userId': 1, 'id': 1, 'title': 'sunt aut facere repellat provident occaecati excepturi optio reprehenderit', 'body': 'quia et suscipit\nsuscipit recusandae consequuntur expedita et cum\nreprehenderit molestiae ut ut quas totam\nnostrum rerum est autem sunt rem eveniet architecto'}
{'userId': 1, 'id': 2, 'title': 'qui est esse', 'body': 'est rerum tempore vitae\nsequi sint nihil reprehenderit dolor beatae ea dolores neque\nfugiat blanditiis voluptate porro vel nihil molestiae ut reiciendis\nqui aperiam non debitis possimus qui neque nisi nulla'}
{'userId': 1, 'id': 3, 'title': 'ea molestias quasi exercitationem repellat qui ipsa sit aut', 'body': 'et iusto sed quo iure\nvoluptatem occaecati omnis eligendi aut ad\nvoluptatem doloribus vel accusantium quis pariatur\nmolestiae porro eius odio et labore et velit aut'}
{'userId': 1, 'id': 4, 'title': 'eum et est occaecati', 'body': 'ullam et saepe reiciendis voluptatem adipisci\nsit amet autem assumenda provi

- **Status Code 200:** Indicates a successful API call.
- **Error Handling:** If the requests package is not installed, an error like `ModuleNotFoundError` will appear. Installing the package and activating the appropriate Python environment resolves this.

### **Example: Making a POST Request**

In [ ]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"

# Data to send with the request
data = {
    "title": "New Post",
    "body": "This is a new post created using the requests module.",
    "userId": 1
}

# Send the POST request
response = requests.post(url, json=data)

# Print status code and the response content
print(f"Status Code: {response.status_code}")

print("Response Content:", response.json())

Status Code: 201
Response Content: {'title': 'New Post', 'body': 'This is a new post created using the requests module.', 'userId': 1, 'id': 101}


- **Status Code 201:** Indicates resource creation was successful.
- The data is sent in JSON format, which is the commonly accepted data interchange format.

## **API Endpoints Explained**
An API Endpoint is the specific URL or address where the API service resides. Think of it as the address of a house you want to visit.

**Example Endpoint:**
> https://api.weather.com/forecast
- When your device requests data (e.g., weather or social media content), it sends the request to the appropriate endpoint.
- The endpoint determines which server and data the client accesses.


## **Types of APIs**
- **REST APIs:** Most common, using HTTP protocols to exchange data in text formats like JSON.
- **SOAP APIs:** Older, more complex protocols (less common in modern applications).
- **GraphQL APIs:** More flexible query language, but less commonly used compared to REST.
**Note:** This lecture focuses primarily on REST APIs due to their widespread use.

## **Practical Insights and Common Use Cases**
- **Weather Apps:** Fetch real-time weather data by making GET requests to a weather API.
- **Social Media:** Apps constantly synchronize with servers via APIs to display up-to-date content.
- **Payment Gateways and Maps:** APIs enable secure payment processing and dynamic map data (e.g., Google Maps).

## **Importance of Error Handling in APIs**
- Always implement exception handling to manage failed API calls gracefully.
- If an API call fails, the response code will indicate the error type (e.g., 404 Not Found, 500 Server Error).
- Proper error handling helps diagnose whether the issue is client-side, network-related, or server-side.

In [3]:
import requests

url = "https://jsonplaceholder.typicode.com/posts"
response = requests.get(url)

# Check if the request was successful
if response.status_code == 200:
    print("Request successful!")

else:
    print(f"Error {response.status_code}: Unable to fetch data.")

Request successful!


## 🔄 Summary of API Workflow
| Step       | Description                                              |
|------------|----------------------------------------------------------|
| Request    | Client sends a request via API to the server.            |
| Processing | Server processes request and prepares a response.        |
| Response   | API returns data or status code back to client.          |